### PR TITLE
Add support for Private & Output to initializer decompose flag

### LINKED ISSUE #2388: When converting from WebGPU expand out initializers
WebGPU requires initiaiizers for a number of storage classes when creating variables. Iniitializers are known to be problematic in some situations, so when converting from WebGPU to another environment , initializers should be unrolled into a creation instruction followed by an immediate store.

### PR DESCRIPTION
**Pull request by zoddicus**
Fixes #2388

### CODE REVIEW

File: source/opt/decompose_initialized_variables_pass.cpp [hunk 1]
```diff
@@ -57,20 +72,41 @@ Pass::Status DecomposeInitializedVariablesPass::Process() {
...
+        uint32_t var_id;
+        uint32_t val_id;
+        std::tie(var_id, val_id) = store_ids;
+        auto* store_inst = new Instruction(
+            context(), SpvOpStore, 0, 0,
+            {{SPV_OPERAND_TYPE_ID, {var_id}}, {SPV_OPERAND_TYPE_ID, {val_id}}});
+        context()->AnalyzeDefUse(store_inst);
+        context()->set_instr_block(store_inst, &*first_block);
+        first_block->AddInstruction(std::unique_ptr<Instruction>(store_inst));
+        if (last_var == first_block->end()) {
```
Comment (s-perron) [inline source/opt/decompose_initialized_variables_pass.cpp]: You can probably simplify a lot of the code related to `last_var` if you changed it from pointing to the last variable to point to the first instruction after the last variable.  That instruction will always exist, and then there is not special case for when there are no variable.  You can just always insert before.

File: source/opt/decompose_initialized_variables_pass.cpp [hunk 2]
```diff
@@ -36,17 +36,32 @@ Pass::Status DecomposeInitializedVariablesPass::Process() {
...
+  for (auto iter = module->types_values_begin();
+       iter != module->types_values_end(); ++iter) {
+    Instruction* inst = &(*iter);
+    if (!HasInitializer(inst)) continue;
+
+    changed = true;
+    auto var_id = inst->result_id();
+    auto val_id = inst->GetOperand(3).words[0];
+    global_stores.push_back(std::make_tuple(var_id, val_id));
+    iter->RemoveOperand(3);
```
Comment (s-perron) [inline source/opt/decompose_initialized_variables_pass.cpp]: Does this mess up the def-use manager?  You remove the operand which is the use of an id.

File: source/opt/decompose_initialized_variables_pass.cpp [hunk 3]
```diff
@@ -57,20 +72,41 @@ Pass::Status DecomposeInitializedVariablesPass::Process() {
       Instruction* store_inst = new Instruction(
           context(), SpvOpStore, 0, 0,
           {{SPV_OPERAND_TYPE_ID, {var_id}}, {SPV_OPERAND_TYPE_ID, {val_id}}});
-      new_stores.push_back(store_inst);
+      function_stores.push_back(store_inst);
       iter->RemoveOperand(3);
```
Comment (s-perron) [inline source/opt/decompose_initialized_variables_pass.cpp]: Does this mess up the def-use manager as well?

Comment (zoddicus) [reply source/opt/decompose_initialized_variables_pass.cpp]: Yup, fixed it.

File: source/opt/decompose_initialized_variables_pass.cpp (hunk 2 above)
Comment (zoddicus) [reply source/opt/decompose_initialized_variables_pass.cpp]: Yup, fixed it.

File: source/opt/decompose_initialized_variables_pass.cpp [hunk 4]
```diff
@@ -34,43 +34,78 @@ bool HasInitializer(Instruction* inst) {
...
-    for (auto iter = block->begin();
-         iter != block->end() && iter->opcode() == SpvOpVariable; ++iter) {
-      last_var = iter;
+    std::vector<Instruction*> function_stores;
+    auto first_block = func->entry().get();
+    inst_iterator insert_point = first_block->begin();
+    for (auto iter = first_block->begin();
+         iter != first_block->end() && iter->opcode() == SpvOpVariable;
+         ++iter) {
+      // For valid SPIRV-V, there is guaranteed to be atleast one instruction
```
Comment (s-perron) [inline source/opt/decompose_initialized_variables_pass.cpp]: "atleast" should be two words.

Comment (zoddicus) [reply source/opt/decompose_initialized_variables_pass.cpp]: Done

**Pull request merged**